### DISCUSSION TITLE
How to use IdentityMappingRUleType

### DISCUSSION THREAD

**Discussion by sinelabore** [FreeOpcUa/python-opcua#1546] [General]
Hi,

I want to call "AddIdentity()" on an object to add a user to a specific role.
The method requires an IdentityCriteriaType parameter if I understand it correctly.
My plan was to use IdentityCriteriaType and set its attributes criteriaType and criteria accordingly.

But the code line
`identityMappingRuleType = ua.IdentityMappingRuleType()`

triggers this error:

`ValueError: 0 is not a valid IdentityCriteriaType`

Which is ok, because criteriaType is an Enum whose first value is 1 (USERNAME). But how to use `ua.IdentityMappingRuleType` without triggering this error immediately before I even have the chance to set the attributes?
 
Is there a short example how to use this Type or do I make something wrong in general?

Peter

**Comment by AndreasHeine** [comment]
https://github.com/FreeOpcUa/opcua-asyncio/blob/90044659adbac85134fdc024b894c6f60135b5e1/asyncua/ua/uaprotocol_auto.py#L1892

https://docs.python.org/3/library/dataclasses.html

the enum is here:

https://github.com/FreeOpcUa/opcua-asyncio/blob/90044659adbac85134fdc024b894c6f60135b5e1/asyncua/ua/uaprotocol_auto.py#L130

which has no member 0!

**Reply by AndreasHeine** [reply]
i assume you should do something like

identityMappingRuleType = ua.IdentityMappingRuleType(ua.IdentityCriteriaType.UserName, "CreteriaString")

**Comment by sinelabore** [comment]
Hi,  your link goes to the async codebase. And there the class IdentityMappingRuleType looks different. 

In the "non async" version the class init code is as follows:

`def __init__(self):
        self.CriteriaType = IdentityCriteriaType(0)
        self.Criteria = None
        self._freeze = True`

Should that be fixed -> **1** instead of 0?
